### PR TITLE
Implement OSS save and load of the sharded state from a single replica

### LINKED ISSUE #8: Optimizer state sharding: collect all states from a single rank
## 🚀 Feature
Make it possible to collect all sharded optimizer states on a single rank.

## Motivation
When saving the current job's state (checkpointing for later job resume), some frameworks expect to be able to fetch the whole training state (including the optimizer's) from a single rank, and they will not trivially be able to query all ranks at that time. 

## Pitch
- Add a "sync state" or similar which is called periodically on all ranks, which brings all the shards to a given rank, in CPU memory space
- Query this reference rank whenever you want to get the full optimizer state for checkpointing

## Alternatives
Modify the frameworks so that they query all replicas, get all the sharded states, bring them back to a single rank or save them directly from there. This is currently not an easy sell

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #8.
Follow up from another PR which got a bit messed up by originating from a fairscale fork

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
I'll change the way the state is built, too many duplications in this PR in between state and param_groups, fixing that

**Comment by blefaudeux** [comment]
works locally on cpu and gpu, but it looks like I'm not there for multi-gpu machines, fixing that but hard to test

**Comment by blefaudeux** [comment]
Finally ready for review I hope, I learnt a thing or two about broadcast and tensor locality limitations (which are not in the doc). Sorry for the botched history (many lint updates needed), I'm setting up my tools for that, coming from FB internal tools

File: fairscale/optim/oss.py [hunk 1]
```diff
@@ -87,17 +97,100 @@ def step(self, closure: Optional[Callable[[], float]] = None) -> Optional[float]
                     dist.broadcast(param, rank, group=self.group)
         return loss
 
-    def state_dict(self) -> dict:
+    def local_state_dict(self) -> dict:
```
Comment (msbaines) [inline fairscale/optim/oss.py:100]: LGTM with one nit: Could you add a load_local_state_dict also.

**Comment by blefaudeux** [comment]
Ah I should have done a squash commit, would have been cleaner. My bad, sorry about that

**Pull request merged**